pkg/crdtest/main_test.go

```diff
@@ -0,0 +1,221 @@
+				if !ok {
+					return
+				}
+				if event.Op&fsnotify.Write == fsnotify.Write || event.Op&fsnotify.Create == fsnotify.Create {
+					if filepath.Ext(event.Name) == ".yaml" || filepath.Ext(event.Name) == ".yml" {
+						runTestOnFile(event.Name)
+					}
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				klog.Errorf("Watcher error: %v", err)
+			}
+		}
+	}()
+
+	err = watcher.Add(watchDir)
+	if err != nil {
+		klog.Fatalf("Failed to add dir to watcher: %v", err)
+	}
+
+	// Run tests on existing files
+	files, err := os.ReadDir(watchDir)
+	if err != nil {
+		klog.Fatalf("Failed to read dir %s: %v", watchDir, err)
+	}
+	for _, file := range files {
+		if !file.IsDir() && (filepath.Ext(file.Name()) == ".yaml" || filepath.Ext(file.Name()) == ".yml") {
+			runTestOnFile(filepath.Join(watchDir, file.Name()))
+		}
+	}
+
+	klog.Infof("Watching for changes in %s...", watchDir)
+	select {} // Block forever
+}
+
+func TestMain(m *testing.M) {
+	klog.InitFlags(nil)
+	flag.Parse()
+
+	klog.Info("Running crdtest")
+
+	scheme := runtime.NewScheme()
+
+	var (
+		restConfig *rest.Config
+		testEnv    *envtest.Environment
+		err        error
+	)
+
+	corev1.AddToScheme(scheme)
+	klog.Info("Added corev1 to scheme")
+
+	v1alpha2.Install(scheme)
+	klog.Info("Added v1alpha2 to scheme")
+
+	if *kubeConfig != "" {
+		klog.Infof("Using kubeConfig=%q", *kubeConfig)
+		restConfig, err = clientcmd.BuildConfigFromFlags("", *kubeConfig)
+		if err != nil {
+			panic(fmt.Sprintf("Failed to get restConfig from BuildConfigFromFlags: %v", err))
+		}
+	} else {
+		klog.Info("Using testenv")
+
+		// The version used here MUST reflect the available versions at
+		// controller-runtime repo: https://raw.githubusercontent.com/kubernetes-sigs/controller-tools/HEAD/envtest-releases.yaml
+		// If the envvar is not passed, the latest GA will be used
+		k8sVersion := os.Getenv("K8S_VERSION")
+
+		var paths []string
+		if *crdDir != "" {
+			paths = []string{*crdDir}
+		} else {
+			paths = []string{
+				filepath.Join("..", "..", "config", "crd", "standard"),
+			}
+		}
+
+		klog.Infof("Paths to CRDs: %v", paths)
+
+		testEnv = &envtest.Environment{
+			Scheme:                      scheme,
+			ErrorIfCRDPathMissing:       true,
+			DownloadBinaryAssets:        true,
+			DownloadBinaryAssetsVersion: k8sVersion,
+			CRDInstallOptions: envtest.CRDInstallOptions{
+				Paths:           paths,
+				CleanUpAfterUse: true,
+			},
+		}
+
+		startTs := time.Now()
+		restConfig, err = testEnv.Start()
+		if err != nil {
+			panic(fmt.Sprintf("Error initializing test environment: %v (took %v)", err, time.Since(startTs)))
+		}
+		klog.Infof("testEnv.Start() took %v", time.Since(startTs))
+	}
+
+	globals.k8sClient, err = client.New(restConfig, client.Options{Scheme: scheme})
+	if err != nil {
+		panic(fmt.Sprintf("Failed to get restConfig from BuildConfigFromFlags: %v", err))
+	}
+
+	if *watchDir != "" {
+		watchAndTest(*watchDir)
+		os.Exit(0)
+	}
+
+	rc := m.Run()
```

Comment:
I expected this path to test all yamls from valid and invalid folders and assure they return expected result. We could then add more test cases e.g. to make sure our CEL works?

Reply:
This does do what you described? The invalid cases expect validation errors (see crd_test.go).

The entry point (main_test.go) has to be named so the "go test -c" will add the right entry point.

Reply:
right I think what confused me is that running
`./crdtest -crdDir config/crd/standard` won't do anything, but I guess that is ok